### ISSUE TITLE
upgrading to newer version of sensu with newer embeded ruby doesn't reinstall plugins

### ISSUE THREAD

**Issue by bchaja** [sensu/sensu-puppet#542]
We are using embedded ruby when installing newer version of sensu with newer embedded ruby the plugins not being reinstalled making them unavailable under upgraded version (also we are using provider sensu_gem). 2 puppet run fixes the issue. All plugins are defined in hiera as:

sensu_gems:
   sensu-plugins-cpu-checks: {}

**Comment by dmizelle** [comment]
We have also run into this issue, and have to solve it by running puppet again. Is there some way to perhaps create a subscribe parameter on each plugin to 'refresh'/reinstall after the version of sensu changes?

If we update sensu-client right now in our environment, all checks go critical/unknown, which makes us get paged.

**Comment by jaxxstorm** [comment]
I don't use this myself, can you provide me some examples of how you're using this?

**Comment by rowleyaj** [comment]
```
  # When updating sensu we found that the sensu_gem installed plugins would take
  # a second puppet run to install. This ensures that when the Sensu package
  # changes, all sensu_gem installed packages will be notified, and then notifies
  # the sensu services (if installed)
  Package['sensu'] ~>
    Package<| provider == 'sensu_gem' |> ~>
      Service['sensu-client', 'sensu-server', 'sensu-api']
```

We added the above to our profile to work around this. There is still a brief period of time where a check may run and error, but it is better than waiting for 30 minutes for the next puppet run.

These kind of dependencies could be captured in the module using the above, there may be others that can be resolved this way too, for example #463

**Comment by glarizza** [comment]
Hopping on this ticket to replicate and see what I can update - is this still a problem for all involved with the recent module changes (presuming so but asking just to be sure).  Does anyone have any code they're using to replicate this issue (I'm presuming install a specific version with all plugins, specify a new version, and then observe)?

**Comment by ghoneycutt** [comment]
@agoddard Any thoughts on how we can reproduce? Should the sensu package itself handle this so gems with the old version are attempted to be reinstalled?

**Comment by glarizza** [comment]
As I understand it, plugins to Sensu's embedded ruby installation are installed via package resources with the `sensu_gem` provider included in this module:

```puppet
package { 'redphone':
  ensure   => 'installed',
  provider => sensu_gem,
}

package { 'sensu-plugins-nginx':
  ensure   => 'installed',
  provider => sensu_gem,
}

package { 'sensu-plugins-disk-checks':
  ensure   => 'installed',
  provider => sensu_gem,
}
```

The `sensu_gem` provider installs into `/opt/sensu/embedded`, and the version that I pulled up has gems installed within `/opt/sensu/embedded/lib/ruby/gems/2.4.0/gems` because the current version of Sensu uses Ruby MRI 2.4 . Problems arise when you upgrade Sensu from an older version with an older Ruby MRI installed - the package drops down a new Ruby MRI version folder and the embedded `gem` binary uses the new folder for plugins/gems.

As stated above, there needs to be a mechanism for when the 'sensu' package changes that the package resources installing plugins into the embedded rubygems path are notified so they can do their installation checks AFTER Sensu is updated.

We can't set an "autonotify" from the plugin resources for the Sensu services because that's implemented in the core Package Type (and while we don't want EVERY package resource to notify the Sensu services Puppet doesn't give us a mechanism to confine it to only those package resources using the `sensu_gem` provider).

Like @rowleyaj said above, you could implement the following within your organization's Sensu Enterprise profile using collectors and the resource chain syntax:

```puppet
Package['sensu'] ~>
Package<| provider == 'sensu_gem' |> ~>
Service['sensu-client', 'sensu-server', 'sensu-api']
```

That code would setup notify dependencies on the 'sensu' package to notify all Package resources using the `sensu_gem` provider, and then setup notify dependencies on all those package resources to notify the Sensu services if changes were made. This works, but is that something we want to drop right into the main Sensu module?  The biggest concern would be whether this bit of magic would cause unexpected circular dependencies, but I feel like it's confined pretty well (limited to 4 specific resources that are declared within this module and any plugins declared using the provider within this module).

I'm going to replicate this and play around with it a bit more, but I'm leaning toward just accepting that code into the module since this is fundamentally a dependency issue and that's the "least magical" solution that actually works.

**Comment by ghoneycutt** [comment]
+1 to this approach. Excited to hear how it works out with your functional testing.

**Comment by ghoneycutt** [comment]
Great work!

Could we have a simple shell script that does what you are doing and tests that output of the last puppet apply? That would give a quick way to functionally test this in the future to ensure we don't have regressions.

**Comment by agoddard** [comment]
Sorry, I thought I’d commented the other day but I guess I’d left it w/out submitting it - the way we’ve seen this done before is to have the plugin installation tasks run after the package installer, wherein it checks the embedded gem’s installed gems, and any that are missing are installed. This works on the first install and if gems are uninstalled manually on the box, as well as if the embedded ruby is upgraded, completely separate from any interaction with the package itself. Would this approach work with Puppet? Basically it would just check /opt/sensu/embedded/bin/gem to see if a gem is installed, and if it’s missing, it will resintall it.

**Comment by glarizza** [comment]
@ghoneycutt The problem is that because those service resources fail to start on the downgrade the exit code reflects that. I can write up the script so a human can review it for regressions, but automating that check will be hard because of the exit code.  A Beaker test might be better if you want to automate that check, and I'll look at that next

**Comment by ghoneycutt** [comment]
@glarizza the existing beaker tests are very outdated and should be removed, don't use those.

**Comment by glarizza** [comment]
@ghoneycutt Good to know, I'll focus my efforts on the script

**Comment by glarizza** [comment]
@agoddard What you wrote is the same angle I ended up taking, so that works out.  Can you tell me if the sensu-enterprise service (or anything on the Enterprise side) needs to be refreshed on plugin changes or if the client/server/api are the only services that would need refreshed?  Basically, do we have to do anything extra on the enterprise side that I haven't already done?

**Comment by glarizza** [comment]
@ghoneycutt Validation script was added in ebeeeef and is on the PR

**Comment by glarizza** [comment]
Tests are failing on some rspec-puppet tests because of the last line:

```puppet
Package['sensu'] ~>
Package<| provider == 'sensu_gem' |> ~>
Service['sensu-client', 'sensu-server', 'sensu-api']
```

On cases where the api or client hasn't been installed there won't be any resources for the `sensu-api` or `sensu-client` service.  In its place I can do something like this:

```puppet
~> Service<| title == 'sensu-client' or title == 'sensu-server' or title == 'sensu-api'|>
```

That collector will discover the resources if they're there and setup the dependencies, or exit silently if they're not in the catalog.  It's not the prettiest, but it will do.

**Comment by ghoneycutt** [comment]
Released in v2.46.0
